### ISSUE TITLE
[Bug]: `make` can't find utilities while compiling gem `ffi`

### ISSUE THREAD

**Issue by gouravkhunger** [termux/termux-packages#11501]
### Problem description

I am compiling `ruby` and `make` using a custom prefix. The bootstraps are working correctly and I can see the binaries work

<img width="546" alt="image" src="https://user-images.githubusercontent.com/46792249/183292068-292e9a06-9890-4a90-b8ff-f8d8faa0b8be.png">

I can also install some gems without a problem. But when I try to install the `ffi` gem, `make` doesn't seems to be able to find core utilities like `ls`, `sh`, `rm`, etc.

Here's the output log:

```
current directory: /data/data/xyz.jekyllex/files/usr/lib/ruby/gems/3.1.0/gems/ffi-1.15.5/ext/ffi_c
/data/data/xyz.jekyllex/files/usr/bin/ruby -I /data/data/xyz.jekyllex/files/usr/lib/ruby/3.1.0 -r ./siteconf20220807-4690-m3xlro.rb extconf.rb
checking for ffi_prep_closure_loc() in -lffi... yes
checking for ffi_prep_cif_var()... yes
checking for ffi_raw_call()... yes
checking for ffi_prep_raw_closure()... yes
checking for whether -pthread is accepted as LDFLAGS... yes
creating extconf.h
creating Makefile

current directory: /data/data/xyz.jekyllex/files/usr/lib/ruby/gems/3.1.0/gems/ffi-1.15.5/ext/ffi_c
make DESTDIR\= clean
make: rm: No such file or directory
make: [Makefile:193: clean-static] Error 127 (ignored)
make: /data/data/xyz.jekyllex/files/usr/bin/sh: No such file or directory
make: [Makefile:176: clean] Error 127 (ignored)

current directory: /data/data/xyz.jekyllex/files/usr/lib/ruby/gems/3.1.0/gems/ffi-1.15.5/ext/ffi_c
make DESTDIR\=
make: echo: No such file or directory
make: *** [Makefile:245: AbstractMemory.o] Error 127

make failed, exit code 2
```

I found that `ls`, `rm`, `mv`, those basic features aren't symlinked:

<img width="442" alt="image" src="https://user-images.githubusercontent.com/46792249/183292189-a7379c3a-4052-4cb6-9614-847bf1343689.png">

Even when I have `dash` and `coreutils` in my custom [`build-bootstraps.sh`](https://github.com/jekyllex/ruby-android/blob/3d506affba2e7c8bc457493179c6eb78f238a0b3/build-bootstraps.sh#L395-L401) file:

```shell
PACKAGES+=("dash")
PACKAGES+=("make")
PACKAGES+=("ruby")
PACKAGES+=("libllvm")
PACKAGES+=("binutils")
PACKAGES+=("coreutils")
PACKAGES+=("pkg-config")
```

The built `SYMLINKS.txt` also has them symlinked to `coreutils`:

<img width="277" alt="image" src="https://user-images.githubusercontent.com/46792249/183292392-1362756f-c0dd-45db-a560-5a9d35d24a11.png">

(`./bin/sh` is also symlinked to `dash`)

What could be the potential problem?

### What steps will reproduce the bug?

- Use custom `$PREFIX` for building bootstrap with the packages listed above.
- After, boostrap install process, `coreutils` won't work even when symlinks are there.

### System information

I don't use `termux-info` in my bootstrap. But I am testing on an `aarch64` emulator if that matters.
I generated bootstraps through [GitHub Actions](https://github.com/jekyllex/ruby-android/actions/runs/2811975846).

**Comment by gouravkhunger** [comment]
Actually, installing `ffi` using `gem install ffi` works on the official Termux app. Just make sure to install `binutils` besides `make` and `ruby`.

I can't seem to get it working on custom `$PREFIX` even with required packages installed.

**Comment by Grimler91** [comment]
What's the output of `file $PREFIX/bin/coreutils`?

strace should be able to tell what file it is that is actually missing: `strace -o ls.log ls`

**Comment by gouravkhunger** [comment]
@Grimler91 The package that has `file` and `strace` binaries is probably not there in my bootstrap:

<img width="374" alt="image" src="https://user-images.githubusercontent.com/46792249/183293724-c523445c-e98e-4449-ab48-3a40aeb6cc4b.png">

**Comment by Grimler91** [comment]
Well, would help if you build and install them then

**Comment by gouravkhunger** [comment]
@Grimler91 I tried building and installing a bootstrap for package [`file`](https://github.com/termux/termux-packages/tree/master/packages/file) and [`strace`](https://github.com/termux/termux-packages/tree/master/packages/strace), still doesn't work.

<img width="465" alt="image" src="https://user-images.githubusercontent.com/46792249/183299011-7621bf97-fecb-46ce-a7f9-ba7f79c50eb0.png">

Am I compiling the wrong packages?

**Comment by licy183** [comment]
Hello, it seems that you are using the shell from `/system/bin/toybox`. Have you set your PATH environment variable properly? Try seting `PATH=YOUR_CUSTOM_PREFIX`.

**Comment by gouravkhunger** [comment]
I think the `coreutils` binary was never moved to the `bin` folder during bootstrap installation process.

<img width="387" alt="image" src="https://user-images.githubusercontent.com/46792249/183299830-7ad18e68-1236-4d30-a3b4-2bfaae9a30f8.png">

But when I extract the zip I can see the `coreutils` binary. Some problem is there during inflating the file to desired location ig.

@licy183 where do I set the `PATH`? On the android app or during some build step 🤔 

Thanks for the help 🙌

**Comment by truboxl** [comment]
Please run the command `env` in the app and paste the output here

**Comment by chyn0907** [comment]
How do you get this shell? Are you using `adb shell`? If you are using adb, to test the bootstrap, you need root permission and some command like the following.

```
BASE_DIR="/data/data/xyz.jekyllex/files"
su $(stat -c %u $BASE_DIR)
export PREFIX="$BASE_DIR/usr"
export HOME="$BASE_DIR/home"
export TMPDIR="$BASE_DIR/usr/tmp"
export LD_LIBRARY_PATH="$PREFIX/lib:$LD_LIBRARY_PATH"
export PATH="$PREFIX/bin:$PATH"
export TZ="UTC"
export LANG="en_US.UTF-8"
export SHELL="$PREFIX/bin/bash"
exec "$SHELL"
```

Then you will get a shell like Termux.

**Comment by gouravkhunger** [comment]
@truboxl As I told, my bootstrap only has specific binaries needed for my use case. The `env` binary is not there, could you please tell the package to be included to have the `env` binary setup?

@chyn0907 I am using an emulator from Android studio. I cloned the termux app repository and changed the package instances as per my need (just for testing my bootstraps, won't use termux app in the final app as it is).

<img width="1436" alt="image" src="https://user-images.githubusercontent.com/46792249/183303926-3a613c4a-22d9-4b56-b558-b19aa7d95510.png">

**Comment by licy183** [comment]
`env` is a sym-link to `coreutils` in Termux, but you could also use the one from toybox like `/system/bin/env`.
I think it is abnormal that the shell from the Android application starts with `/system/bin/sh` rather than `$PREFIX/bin/sh`, which is the default behavior of the failsafe session, but I have no idea why this happens. So, if it is convenient, would you please provide the source code or provide a patch file between the origin termux source code from your fork?

**Comment by gouravkhunger** [comment]
@licy183 I just changed `com.termux` instances to `xyz.jekyllex` across the source files and the package folder names. But I did this some 3-4 months ago so it might not be up to date as per the new app. The patch would become so large, but basically there's change of package name only. I haven't stored it on any remote repo.

Regarding the `env` symlink - that's the problem. I have `coreutils` there in my bootstrap. But it is [not being installed correctly](https://github.com/termux/termux-packages/issues/11501#issuecomment-1207437538) when I do a fresh install on the emulator. When I extract the files from bootstrap zip, I see `coreutils` binary as well as symlinks in `SYMLINKS.txt` file. But when I explore the content of emulator, all other binaries are there (the reason why `ruby` and `make` work) except `coreutils` and `dash`.

**Comment by gouravkhunger** [comment]
I very very sorry guys 😔 

It was all due to Gradle using cached older bootstrap zip (which didn't have `dash` and `coreutils`) rather than ones with latest packages. When I first got the error from `make`, about not being able to find utilities like `ls`, `rm` and `sh`, I knew that I needed `dash` and `coreutils`. 

I rebuilt the bootstrap, tried different things but still got the same errors. But then I tried building bootstrap for just `coreutils` and after exploring device file manager I still round `ruby`, and older binaries which made me skeptical of the caches.

I force deleted app from the emulator, deleted all the caches, `.gradle`, `.idea` folder. Re-added bootstraps to `cpp` folder. Now things worked!

<img width="393" alt="image" src="https://user-images.githubusercontent.com/46792249/183306178-026bbd88-7ca8-40e8-8480-c4a1dbfac43a.png">

Again, I am very sorry for wasting your time. I was stuck in this for 2+ days so thought of getting some expert help 😓 

Thank you for your precious time, everyone.
